[GSoC] `mrv_leadterm` changes

0sidharth opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #6052
Closes #6350

#### Brief description of what is fixed or changed
`calculate_series` function of `gruntz` has been removed. Some changes across many other files for better `logx` handling.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* series
  * Removed `calculate_series` function from `gruntz`
<!-- END RELEASE NOTES -->

0sidharth committed:
Fixed logx not being passed bug in Add, removed an exception from Mul._eval_as_leading_term

0sidharth committed:
Fixed improper handling of signs and logx in polygamma and Ei

0sidharth committed:
Better handling of logx and Order addition in log._eval_as_leading_term

0sidharth committed:
Removed calculate_series function from gruntz

0sidharth committed:
Fixing failing test cases

jksuom commented:
We should probably deprecate `calculate_series` before removing it. Somebody may be using it.

jksuom commented on sympy/functions/elementary/exponential.py:
```diff
@@ -991,7 +993,10 @@ def coeff_exp(term, x):
+            if (log(a) + b*log(x)) == self:
+                return log(a) + b*logx
+            else:
+                return log(a) + b*logx + Order(x**(n + 1), x)
```
Why `n + 1`?

0sidharth replied:
That's a mistake, that was actually masking another bug (value of `n` was changing in the method), which I have changed now.

0sidharth committed:
Add back calculate_series with Deprecation warning

Fixed bug in SingularityFunction._eval_nseries and doctest rewritten

0sidharth committed:
Added besseli leading term method

0sidharth committed:
Fixed bug related to parameter n in log._eval_nseries

0sidharth committed:
Remove use of compute_leading_term (now gives deprecation warning)

0sidharth committed:
Fixed Failing tests

atan._eval_as_leading_term zoo case handled better
Ei._eval_as_leading_term fixed

jksuom commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -2662,7 +2662,8 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
         if x0.is_zero:
             return arg.as_leading_term(x)
         if x0 is S.ComplexInfinity:
-            return acot(1/arg)._eval_as_leading_term(x, cdir=cdir)
```
Does this give a wrong result?

jksuom commented on sympy/functions/special/gamma_functions.py:
```diff
@@ -1057,6 +1057,19 @@ def fdiff(self, argindex=1):
+                r = r.replace(log(c*x**ex), log(c) + ex*logx)
+                return r
+            return z*(-log(z))
+        return log(gamma(z))._eval_as_leading_term(x, logx=logx, cdir=cdir)
```
This could probably be `return loggamma(z0)` as `gamma` is nonzero everywhere.

jksuom replied:
Strictly speaking, we should also handle the poles of `gamma(z)` at `z = 0, -1, -2, ...` but that does not seem necessary for now.

0sidharth replied:
It can be changed to that when `logx` is None, otherwise it may give unexpected results

oscargus committed:
Merge branch 'master' into mrv_leadterm_changes

oscargus commented on sympy/functions/special/bessel.py:
```diff
@@ -496,7 +496,19 @@ def _eval_is_extended_real(self):
+        import sage.all as sage
+        return sage.bessel_I(self.args[0]._sage_(), self.args[1]._sage_())
 
+      
```
```suggestion

```

oscargus committed:
Apply suggestions from code review

oscargus commented on sympy/series/gruntz.py:
```diff
@@ -608,6 +604,8 @@ def rewrite(e, Omega, x, wsym):
     Returns the rewritten e in terms of w and log(w). See test_rewrite1()
     for examples and correct results.
     """
+    from sympy import AccumBounds, ilcm
```
```suggestion
    from sympy import AccumBounds
```

oscargus commented on sympy/series/gruntz.py:
```diff
@@ -608,6 +604,8 @@ def rewrite(e, Omega, x, wsym):
     for examples and correct results.
     """
+    from sympy import AccumBounds, ilcm
+    from sympy.core.traversal import bottom_up
```
```suggestion
```

oscargus committed:
Apply suggestions from code review

0sidharth commented:
I am adding below the latest diff that I was trying around for further development. I have tried a few times but have been unable to get all the tests passing, so feel free take over.
```diff
diff --git a/sympy/core/add.py b/sympy/core/add.py
index 3ae5b0e6c6..4734be47b9 100644
--- a/sympy/core/add.py
+++ b/sympy/core/add.py
@@ -983,7 +983,10 @@ def as_real_imag(self, deep=True, **hints):
     def _eval_as_leading_term(self, x, logx=None, cdir=0):
         from sympy import expand_mul, Order, Piecewise, piecewise_fold, log
 
-        old = self
+        o = self.getO()
+        if o is None:
+            o = Order(0)
+        old = self.removeO()
 
         if old.has(Piecewise):
             old = piecewise_fold(old)
@@ -1037,7 +1040,7 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
             return res.as_leading_term(x, logx=logx, cdir=cdir)
 
         elif new_expr is S.NaN:
-            return old.func._from_args(infinite)
+            return old.func._from_args(infinite) + o
 
         else:
             return new_expr
diff --git a/sympy/core/mul.py b/sympy/core/mul.py
index bbe1e26c0a..625f4aaae1 100644
--- a/sympy/core/mul.py
+++ b/sympy/core/mul.py
@@ -1908,7 +1908,7 @@ def coeff_exp(term, x):
             lt = term.as_coeff_exponent(x)
             if lt[0].has(x):
                 try:
-                    lt = term.leadterm(x)
+                    lt = term.leadterm(x, logx=logx, cdir=cdir)
                 except ValueError:
                     return term, S.Zero
             return lt
@@ -1917,7 +1917,7 @@ def coeff_exp(term, x):
 
         try:
             for t in self.args:
-                coeff, exp = t.leadterm(x, logx=logx)
+                coeff, exp = t.leadterm(x, logx=logx, cdir=cdir)
                 if not coeff.has(x):
                     ords.append((t, exp))
                 else:
diff --git a/sympy/core/power.py b/sympy/core/power.py
index 49070d58bf..5123881ebf 100644
--- a/sympy/core/power.py
+++ b/sympy/core/power.py
@@ -1627,6 +1627,8 @@ def _eval_nseries(self, x, n, logx, cdir=0):
             b = b._eval_nseries(x, n=max(2, n), logx=logx, cdir=cdir).removeO()
             if b.has(nan, zoo):
                 raise NotImplementedError()
+            if b.has(log):
+                return self
             _, m = b.leadterm(x)
 
         if e.has(log):
@@ -1673,7 +1675,7 @@ def mul(d1, d2):
             return res
 
         try:
-            _, d = g.leadterm(x)
+            _, d = g.leadterm(x, logx=logx)
         except (ValueError, NotImplementedError):
             if limit(g/x**maxpow, x, 0) == 0:
                 # g has higher order zero
@@ -1740,7 +1742,10 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
                 pass
             return lt
         else:
-            f = b.as_leading_term(x, logx=logx, cdir=cdir)
+            try:
+                f = b.as_leading_term(x, logx=logx, cdir=cdir)
+            except PoleError:
+                return self
             if (not e.is_integer and f.is_constant() and f.is_real
                 and f.is_negative and im((b - f).dir(x, cdir)).is_negative):
                 return self.func(f, e) * exp(-2 * e * S.Pi * I)
diff --git a/sympy/functions/elementary/exponential.py b/sympy/functions/elementary/exponential.py
index 72e5dc6e27..6fcd055130 100644
--- a/sympy/functions/elementary/exponential.py
+++ b/sympy/functions/elementary/exponential.py
@@ -983,14 +983,14 @@ def coeff_exp(term, x):
 
         # TODO new and probably slow
         try:
-            a, b = arg.leadterm(x)
+            a, b = arg.leadterm(x, logx=logx)
             s = arg.nseries(x, n=n+b, logx=logx)
         except (ValueError, NotImplementedError, PoleError):
             s = arg.nseries(x, n=n, logx=logx)
             while s.is_Order:
                 n += 1
                 s = arg.nseries(x, n=n, logx=logx)
-        a, b = s.removeO().leadterm(x)
+        a, b = s.removeO().leadterm(x, logx=logx)
         p = cancel(s/(a*x**b) - 1).expand().powsimp()
         if p.has(exp):
             p = logcombine(p)
@@ -1041,11 +1041,11 @@ def mul(d1, d2):
     def _eval_as_leading_term(self, x, logx=None, cdir=0):
         from sympy import I, im, re
         arg0 = self.args[0].together()
-        arg = arg0.as_leading_term(x, cdir=cdir)
+        arg = arg0.as_leading_term(x, logx=logx, cdir=cdir)
 
         if logx is not None:
             try:
-                c, e = arg.leadterm(x, cdir=cdir)
+                c, e = arg.leadterm(x, logx=logx, cdir=cdir)
             except ValueError:
                 # arg must be approaching infinity, so should not be replaced
                 # by a constant
@@ -1055,16 +1055,17 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
 
         x0 = arg0.subs(x, 0)
         if (x0 is S.NaN and logx is None):
-            x0 = arg.limit(x, 0, dir='-' if re(cdir).is_negative else '+')
-        if x0 in (S.NegativeInfinity, S.Infinity):
-            raise PoleError("Cannot expand %s around 0" % (self))
+            x0 = arg0.limit(x, 0, dir='-' if re(cdir).is_negative else '+')
         if x0 == 1:
-            return (arg0 - S.One).as_leading_term(x)
+            return (arg0 - S.One).as_leading_term(x, logx=logx, cdir=cdir)
+        if x0.is_infinite:
+            raise PoleError("Cannot expand %s around 0" % (self))
         if cdir != 0:
             cdir = arg0.dir(x, cdir)
         if x0.is_real and x0.is_negative and im(cdir).is_negative:
             return self.func(x0) - 2*I*S.Pi
-        return self.func(arg)
+        c, e = arg.leadterm(x, logx=logx, cdir=cdir)
+        return log(c) if e == 0 else e*log(x)
 
 
 class LambertW(Function):
diff --git a/sympy/functions/special/error_functions.py b/sympy/functions/special/error_functions.py
index e9f6742d9e..b5ea14f763 100644
--- a/sympy/functions/special/error_functions.py
+++ b/sympy/functions/special/error_functions.py
@@ -1210,9 +1210,10 @@ def _eval_rewrite_as_tractable(self, z, limitvar=None, **kwargs):
     def _eval_as_leading_term(self, x, logx=None, cdir=0):
         from sympy import re
         x0 = self.args[0].limit(x, 0)
+        arg = self.args[0].as_leading_term(x, cdir=cdir)
+        cdir = arg.dir(x, cdir)
         if x0.is_zero:
-            f = self._eval_rewrite_as_Si(*self.args)
-            return re(f._eval_as_leading_term(x, logx=logx, cdir=cdir))
+            return log(-arg) + S.EulerGamma if re(cdir).is_negative else log(arg)
         return super()._eval_as_leading_term(x, logx=logx, cdir=cdir)
 
     def _eval_nseries(self, x, n, logx, cdir=0):
@@ -2703,7 +2704,7 @@ class _eis(Function):
     def _eval_aseries(self, n, args0, x, logx):
         from sympy import Order
         if args0[0] != S.Infinity:
-            return super(_erfs, self)._eval_aseries(n, args0, x, logx)
+            return super(_eis, self)._eval_aseries(n, args0, x, logx)
 
         z = self.args[0]
         l = [ factorial(k) * (1/z)**(k + 1) for k in range(0, n) ]
diff --git a/sympy/functions/special/gamma_functions.py b/sympy/functions/special/gamma_functions.py
index 2251a198ec..9db78c9b55 100644
--- a/sympy/functions/special/gamma_functions.py
+++ b/sympy/functions/special/gamma_functions.py
@@ -197,6 +197,16 @@ def _eval_nseries(self, x, n, logx, cdir=0):
         t = self.args[0] - x0
         return (self.func(t + 1)/rf(self.args[0], -x0 + 1))._eval_nseries(x, n, logx)
 
+    def _eval_aseries(self, n, args0, x, logx):
+        from sympy import Order
+        z = self.args[0].as_leading_term(x)
+        r = loggamma(z)._eval_aseries(n + 1, args0, x, logx=None).removeO()
+        return exp(r)
+
+    def _sage_(self):
+        import sage.all as sage
+        return sage.gamma(self.args[0]._sage_())
+
     def _eval_as_leading_term(self, x, logx=None, cdir=0):
         from sympy import PoleError
         arg = self.args[0]
@@ -1026,7 +1036,8 @@ def _eval_aseries(self, n, args0, x, logx):
         else:
             o = Order(1/z**n, x)
         # It is very inefficient to first add the order and then do the nseries
-        return (r + Add(*l))._eval_nseries(x, n, logx) + o
+        return (r._eval_nseries(x, n, None)\
+                + Add(*l)._eval_nseries(x, n, None) + o)
 
     def _eval_rewrite_as_intractable(self, z, **kwargs):
         return log(gamma(z))
@@ -1060,7 +1071,9 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
                 r = r.replace(log(c*x**ex), log(c) + ex*logx)
                 return r
             return z*(-log(z))
-        return log(gamma(z))._eval_as_leading_term(x, logx=logx, cdir=cdir)
+        if z0.is_zero:
+            return log(gamma(z))._eval_as_leading_term(x, logx=logx, cdir=cdir)
+        return loggamma(z0)
 
     def _sage_(self):
         import sage.all as sage
diff --git a/sympy/series/gruntz.py b/sympy/series/gruntz.py
index 1f99851883..d1c42b2023 100644
--- a/sympy/series/gruntz.py
+++ b/sympy/series/gruntz.py
@@ -539,7 +539,8 @@ def mrv_leadterm(e, x):
         ex = f.as_base_exp()[1]
         lt = (base[0]**ex, base[1]*ex)
         if lt[0].has(w):
-            series = f.series(w, n=1, logx=logw).removeO()
+            f = f.rewrite('tractable', deep=True, limitvar=w)
+            series = f.nseries(w, n=1, logx=logw).removeO()
             lt = series.leadterm(w, logx=logw)
     return (lt[0].subs(log(w), logw), lt[1])
 
@@ -613,6 +614,7 @@ def rewrite(e, Omega, x, wsym):
     # make sure we know the sign of each exp() term; after the loop,
     # g is going to be the "w" - the simplest one in the mrv set
     for g, _ in Omega:
+        from sympy import sympify
         sig = sign(g.exp, x)
         if sig != 1 and sig != -1 and not sig.has(AccumBounds):
             raise NotImplementedError('Result depends on the sign of %s' % sig)
diff --git a/sympy/series/tests/test_gruntz.py b/sympy/series/tests/test_gruntz.py
index b89cf8eafe..97246f79fa 100644
--- a/sympy/series/tests/test_gruntz.py
+++ b/sympy/series/tests/test_gruntz.py
@@ -391,6 +391,7 @@ def test_MrvTestCase_page47_ex3_21():
 
 
 def test_I():
+    from sympy.functions import sign
     y = Symbol("y")
     assert gruntz(I*x, x, oo) == I*oo
     assert gruntz(y*I*x, x, oo) == y*I*oo
diff --git a/sympy/series/tests/test_limits.py b/sympy/series/tests/test_limits.py
index 72842c262c..7ca7870a2b 100644
--- a/sympy/series/tests/test_limits.py
+++ b/sympy/series/tests/test_limits.py
@@ -113,7 +113,7 @@ def test_log():
     expr = tau**2*((tau - 1)*(tau + 1)*log(x + 1)/(tau**2 + 1)**2 + 1/((tau**2\
             + 1)*(x + 1)) - (-2*tau*atan(x/tau) + (tau**2/2 - 1/2)*log(tau**2\
             + x**2))/(tau**2 + 1)**2)
-    assert limit(expr, x, oo) == pi*tau**3/(tau**2 + 1)**2
+    assert limit(expr, x, oo).together() == pi*tau**3/(tau**2 + 1)**2
 
 
 def test_piecewise():
diff --git a/sympy/series/tests/test_order.py b/sympy/series/tests/test_order.py
index fd3c4f7f6b..0044267ef5 100644
--- a/sympy/series/tests/test_order.py
+++ b/sympy/series/tests/test_order.py
@@ -157,7 +157,7 @@ def test_add_1():
 def test_ln_args():
     assert O(log(x)) + O(log(2*x)) == O(log(x))
     assert O(log(x)) + O(log(x**3)) == O(log(x))
-    assert O(log(x*y)) + O(log(x) + log(y)) == O(log(x*y))
+    assert O(log(x*y)) + O(log(x) + log(y)) == O(log(x) + log(y))
 
 
 def test_multivar_0():
```

oscargus commented:
Thanks! I basically just wanted to solve the merge conflicts to see the errors. Turned out to be not possible through the standard suggestion interface...
